### ISSUE TITLE
aws_autoscaling_group.asg: aws_autoscaling_group.asg: diffs didn't match during apply. This is a bug with Terraform and should be reported as a GitHub Issue.

### ISSUE THREAD

**Issue by sandeep000** [terraform-providers/terraform-provider-aws#1990]
Hi there,

I am getting following error while applying the terraform script. Please help me resolve this

Error: "aws_autoscaling_group.asg: aws_autoscaling_group.asg: diffs didn't match during apply. This is a bug with Terraform and should be reported as a GitHub Issue."

### Terraform Version
Terraform Version: 0.10.7

### Affected Resource(s)
Resource ID: aws_autoscaling_group.asg

### Terraform Configuration Files
```hcl
# Specify the provider and access details
provider "aws" {
  region = "${var.aws_region}"
  profile = "${var.aws_profile}"
}

# Create a VPC to launch our instances into
resource "aws_vpc" "vpc" {
  cidr_block = "10.0.0.0/16"
}

# Create an internet gateway to give our subnet access to the outside world
resource "aws_internet_gateway" "internet_gateway" {
  vpc_id = "${aws_vpc.vpc.id}"
  tags {
  	Name = "terraform generated internet_gateway"
  }
}

resource "aws_route_table" public {
  vpc_id = "${aws_vpc.vpc.id}"
  route {
  	cidr_block = "0.0.0.0/0"
	gateway_id             = "${aws_internet_gateway.internet_gateway.id}"
  }
  tags {
  	Name = "terraform generated public route table"
  }
}



resource "aws_route_table" "private" {
  vpc_id                  = "${aws_vpc.vpc.id}"
  tags {
  	Name = "terraform generated private route table"
  }
}

# Create a subnet to launch our instances into
resource "aws_subnet" "public" {
  vpc_id                  = "${aws_vpc.vpc.id}"
  cidr_block              = "10.0.1.0/24"
  map_public_ip_on_launch = true
  tags {
	Name = "terraform generated public_subnet"
  }
}

resource "aws_route_table_association" "route_table_assocaition" {
  subnet_id = "${aws_subnet.public.id}"
  route_table_id = "${aws_route_table.public.id}"
}


# Our default security group to access
# the instances over SSH and HTTP
resource "aws_security_group" "default" {
  name        = "terraform_instance_sg"
  description = "Used in the terraform"
  vpc_id      = "${aws_vpc.vpc.id}"

  # SSH access from anywhere
  ingress {
    from_port   = 22
    to_port     = 22
    protocol    = "tcp"
    cidr_blocks = ["0.0.0.0/0"]
  }

  # HTTP access from anywhere
  ingress {
    from_port   = 80
    to_port     = 80
    protocol    = "tcp"
    cidr_blocks = ["0.0.0.0/0"]
  }

  # outbound internet access
  egress {
    from_port   = 0
    to_port     = 0
    protocol    = "-1"
    cidr_blocks = ["0.0.0.0/0"]
  }

  lifecycle {
    create_before_destroy = true
  }
}

resource "aws_key_pair" "auth" {
  key_name   = "${var.key_name}"
  public_key = "${file(var.public_key_path)}"
}


resource "aws_instance" "web" {
  # The connection block tells our provisioner how to
  # communicate with the resource (instance)
  connection {
    # The default username for our AMI
    user = "ec2-user"

    # The connection will use the local SSH agent for authentication.
  }

  instance_type = "t2.micro"

  # Lookup the correct AMI based on the region
  # we specified
  ami = "${lookup(var.aws_amis, var.aws_region)}"

  # The name of our SSH keypair we created above.
  key_name = "${aws_key_pair.auth.id}"

  # Our Security group to allow HTTP and SSH access
  vpc_security_group_ids = ["${aws_security_group.default.id}"]

  # We're going to launch into the same subnet as our ELB. In a production
  # environment it's more common to have a separate private subnet for
  # backend instances.
  subnet_id = "${aws_subnet.public.id}"
  #count = 2

  # We run a remote provisioner on the instance after creating it.
  # modify the directory permissions in a remote-exec
  provisioner "remote-exec" {
    inline = [
      "sudo yum -y update",
      "sudo yum -y install httpd",
      "sudo systemctl start httpd",
      "sudo systemctl enable httpd",
      "sudo chown -R ec2-user /var/www/html",
    ]
  }

  

  

  provisioner "file" {
    source      = "/home/terraform/example2/index.html"
    destination = "/var/www/html/index.html"
    connection {
    type     = "ssh"
    user     = "ec2-user"
    }
  
  }
}

# now we start preparing for the load balancer and ASG

# Create a subnet to launch our instances into ELB and ASG
resource "aws_subnet" "public2" {
  vpc_id                  = "${aws_vpc.vpc.id}"
  cidr_block              = "10.0.2.0/24"
  map_public_ip_on_launch = true
  tags {
	Name = "terraform generated public2_subnet"
  }
}

resource "aws_subnet" "public3" {
  vpc_id                  = "${aws_vpc.vpc.id}"
  cidr_block              = "10.0.3.0/24"
  map_public_ip_on_launch = true
  tags {
	Name = "terraform generated public3_subnet"
  }
}

resource "aws_route_table_association" "route_table_assocaition_public2" {
  subnet_id = "${aws_subnet.public2.id}"
  route_table_id = "${aws_route_table.public.id}"
}

resource "aws_route_table_association" "route_table_assocaition_public3" {
  subnet_id = "${aws_subnet.public3.id}"
  route_table_id = "${aws_route_table.public.id}"
}

# lets create a security group for the ELB so it is accessible via the web
resource "aws_security_group" "elb" {
  name        = "terraform_elb_sg"
  description = "Used in the terraform"
  vpc_id      = "${aws_vpc.vpc.id}"

  # HTTP access from anywhere
  ingress {
    from_port   = 80
    to_port     = 80
    protocol    = "tcp"
    cidr_blocks = ["0.0.0.0/0"]
  }

  # outbound internet access
  egress {
    from_port   = 0
    to_port     = 0
    protocol    = "-1"
    cidr_blocks = ["0.0.0.0/0"]
  }
}

#data "aws_availability_zones" "all" {}

resource "aws_elb" "web" {
  name = "terraform-created-elb"

  subnets         = ["${aws_subnet.public2.id}", "${aws_subnet.public3.id}"]
  security_groups = ["${aws_security_group.elb.id}"]
  #availability_zones = ["${data.aws_availability_zones.all.names}"]
  cross_zone_load_balancing   = true
  idle_timeout                = 400
  connection_draining         = true
  connection_draining_timeout = 400

  listener {
    instance_port     = 80
    instance_protocol = "http"
    lb_port           = 80
    lb_protocol       = "http"
  }

  health_check {
  healthy_threshold = 2
  unhealthy_threshold = 2
  timeout = 3
  target = "HTTP:80/"
  interval = 30
  }
}

# Let us start creating the Auto Scaling

# First we create the AMI

resource "aws_ami_from_instance" "golden" {
  name = "ami-terraform-generated-web"
  source_instance_id = "${aws_instance.web.id}"
}


# Now we create the launch configuration

resource "aws_launch_configuration" "launch_configuration" {
  image_id = "${aws_ami_from_instance.golden.id}"
  instance_type = "t2.micro"
  security_groups = ["${aws_security_group.default.id}"]
  key_name = "${aws_key_pair.auth.id}" 

  
  lifecycle {
    create_before_destroy = true
  }
}


# Now we create the Auto Scaling Group



resource "aws_autoscaling_group" "asg" {
  launch_configuration = "${aws_launch_configuration.launch_configuration.id}"
  availability_zones = ["${var.aws_region}a","${var.aws_region}b"]
  vpc_zone_identifier = ["${aws_subnet.public2.id}","${aws_subnet.public3.id}"]
  min_size = 2
  max_size = 10             
  health_check_grace_period = 300
  desired_capacity          = 4
  force_delete              = true
  

  load_balancers = ["${aws_elb.web.name}"]
  health_check_type = "ELB"
  
  lifecycle {
    create_before_destroy = true
    ignore_changes = ["desired_capacity"]
  }
  
  tag {
    key = "Name"
    value = "terraform-generated-asg"
    propagate_at_launch = true
  }
}


output "elb_dns_name" {
  value = "${aws_elb.web.dns_name}"
}

### Debug Output

Error applying plan:

1 error(s) occurred:

* aws_autoscaling_group.asg: aws_autoscaling_group.asg: diffs didn't match during apply. This is a bug with Terraform and should be reported as a GitHub Issue.

Please include the following information in your report:

    Terraform Version: 0.10.7
    Resource ID: aws_autoscaling_group.asg
    Mismatch reason: attribute mismatch: availability_zones.221770259
    Diff One (usually from plan): *terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff{"metrics_granularity":*terraform.ResourceAttrDiff{Old:"", New:"1Minute", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "availability_zones.2487133097":*terraform.ResourceAttrDiff{Old:"", New:"us-west-2a", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "launch_configuration":*terraform.ResourceAttrDiff{Old:"", New:"${aws_launch_configuration.launch_configuration.id}", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "arn":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "tag.173408575.key":*terraform.ResourceAttrDiff{Old:"", New:"Name", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "tag.173408575.propagate_at_launch":*terraform.ResourceAttrDiff{Old:"", New:"true", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "vpc_zone_identifier.#":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "load_balancers.2554202764":*terraform.ResourceAttrDiff{Old:"", New:"terraform-created-elb", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "health_check_type":*terraform.ResourceAttrDiff{Old:"", New:"ELB", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "availability_zones.221770259":*terraform.ResourceAttrDiff{Old:"", New:"us-west-2b", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "desired_capacity":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "health_check_grace_period":*terraform.ResourceAttrDiff{Old:"", New:"300", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "protect_from_scale_in":*terraform.ResourceAttrDiff{Old:"", New:"false", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "load_balancers.#":*terraform.ResourceAttrDiff{Old:"", New:"1", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "tag.#":*terraform.ResourceAttrDiff{Old:"", New:"1", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "availability_zones.#":*terraform.ResourceAttrDiff{Old:"", New:"2", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "tag.173408575.value":*terraform.ResourceAttrDiff{Old:"", New:"terraform-generated-asg", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "wait_for_capacity_timeout":*terraform.ResourceAttrDiff{Old:"", New:"10m", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "min_size":*terraform.ResourceAttrDiff{Old:"", New:"2", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "force_delete":*terraform.ResourceAttrDiff{Old:"", New:"false", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "max_size":*terraform.ResourceAttrDiff{Old:"", New:"10", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "default_cooldown":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "name":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:true, Sensitive:false, Type:0x0}, "target_group_arns.#":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}}, Destroy:false, DestroyDeposed:false, DestroyTainted:false, Meta:map[string]interface {}(nil)}
    Diff Two (usually from apply): *terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff{"arn":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "tag.173408575.key":*terraform.ResourceAttrDiff{Old:"", New:"Name", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "load_balancers.2554202764":*terraform.ResourceAttrDiff{Old:"", New:"terraform-created-elb", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "protect_from_scale_in":*terraform.ResourceAttrDiff{Old:"", New:"false", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "health_check_type":*terraform.ResourceAttrDiff{Old:"", New:"ELB", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "tag.173408575.propagate_at_launch":*terraform.ResourceAttrDiff{Old:"", New:"true", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "vpc_zone_identifier.#":*terraform.ResourceAttrDiff{Old:"", New:"1", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "load_balancers.#":*terraform.ResourceAttrDiff{Old:"", New:"1", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "wait_for_capacity_timeout":*terraform.ResourceAttrDiff{Old:"", New:"10m", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "name":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:true, Sensitive:false, Type:0x0}, "desired_capacity":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "min_size":*terraform.ResourceAttrDiff{Old:"", New:"2", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "force_delete":*terraform.ResourceAttrDiff{Old:"", New:"false", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "target_group_arns.#":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "tag.#":*terraform.ResourceAttrDiff{Old:"", New:"1", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "tag.173408575.value":*terraform.ResourceAttrDiff{Old:"", New:"terraform-generated-asg", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "vpc_zone_identifier.3465488053":*terraform.ResourceAttrDiff{Old:"", New:"subnet-090d6c6f", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "max_size":*terraform.ResourceAttrDiff{Old:"", New:"10", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "default_cooldown":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "health_check_grace_period":*terraform.ResourceAttrDiff{Old:"", New:"300", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "metrics_granularity":*terraform.ResourceAttrDiff{Old:"", New:"1Minute", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "launch_configuration":*terraform.ResourceAttrDiff{Old:"", New:"terraform-20171020202756231200000001", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}}, Destroy:false, DestroyDeposed:false, DestroyTainted:false, Meta:map[string]interface {}(nil)}

Also include as much context as you can about your config, state, and the steps you performed to trigger this error.


Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.


### Steps to Reproduce
Steps required to reproduce the issue, 
1. `terraform apply`

[variable.tf.txt](https://github.com/terraform-providers/terraform-provider-aws/files/1403254/variable.tf.txt)
[main.tf.txt](https://github.com/terraform-providers/terraform-provider-aws/files/1403255/main.tf.txt)

**Comment by xenoterracide** [comment]
I think #2117 is probably the same issue, terraform 0.10.8, aws 1.2, interesting note, if I create everything including my network settings first, it doesn't happen. and running things twice fixes it.

**Comment by sandeep000** [comment]
Yes, running the script twice fixes this. Thank you!

**Comment by xenoterracide** [comment]
still exists 11.0/1.2

**Comment by xenoterracide** [comment]
same broken behavior ```
C:\Users\xeno\IdeaProjects\dex\etc\terraform\build>terraform -version
Terraform v0.11.1
+ provider.aws v1.5.0
+ provider.template v1.0.0
```

**Comment by xenoterracide** [comment]
Also worth mentioning, this wasn't a problem in 0.9.x

**Comment by bflad** [comment]
This issue was automatically closed by the documentation update in #3897. The configuration fix for `diffs didn't match during apply` and `Mismatch reason: attribute mismatch: availability_zones.XXXXXX` is to remove `availability_zones` configuration if `vpc_zone_identifier` is also specified. If you must specify `availability_zones` due to being used in a module that works with both EC2-Classic and VPC, `availability_zones` can be set to `[]` (an empty list).
